### ISSUE TITLE
map-reduce stack-effect

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2476]
This probably should return 12, but it returns 24.

```factor
2 { 3 3 } [ B dupd * ] [ B + ] map-reduce
```

You can see the issue is it applies the map quotation twice, keeping the result of the first call (6) on the stack for the second call which results in (18) and then the + adds them to get 24.

**Comment by mrjbq7** [comment]
This looks like it fixes it:

```
: map-reduce2 ( ..a seq map-quot: ( ..a elt -- ..b intermediate ) reduce-quot: ( ..b prev intermediate -- ..a next ) -- ..a result )
    [ [ [ first ] keep ] dip [ dip ] keep ] dip '[ swap _ dip swap @ ] 1 each-from ; inline
```

**Comment by mrjbq7** [comment]
And this might fix 2map-reduce:

```factor
: 2map-reduce2 ( ..a seq1 seq2 map-quot: ( ..a elt1 elt2 -- ..b intermediate ) reduce-quot: ( ..b prev intermediate -- ..a next ) -- ..a result )
    [ [ [ [ first ] bi@ ] 2keep ] dip [ 2dip ] keep ] dip
    '[ rot _ dip -rot @ ] 1 2each-from ; inline
```

**Comment by mrjbq7** [comment]
We probably still need to fix ``nmap-reduce``.

**Comment by George66** [comment]
If ...a and ...b have different length, how can we apply the second map-quot? I think, this stack effect is impossible. We can make the following effect
...a seq map-quot: (  ..a elt -- ..a intermediate ) reduce-quot: (..a prev intermediate -- ..a next ) -- ..a result

**Comment by mrjbq7** [comment]
That appears to be correct, I want to think about it a little more, but I suspect you are right.

We would want to leave the ``..a`` to indicate that you can dip below the stack, but consistently.

**Comment by mrjbq7** [comment]
Fixed in 5b1c814d84714b0c94c5a7b4aeb3f44d2a4d5998.
